cachi2/core/package_managers/pip.py

```diff
@@ -1734,78 +1737,70 @@ def _process_package_distributions(
+        dpi = DistributionPackageInfo(
             name,
             version,
-            "sdist",
-            pip_deps_dir.join_within_root(p.filename).path,
-            p.url,
-            p.is_yanked,
+            cast(str, package.package_type),
+            pip_deps_dir.join_within_root(package.filename).path,
+            package.url,
+            package.is_yanked,
+            pypi_checksums,
+            user_checksums,
         )
-        for p in packages
-        if p.version
-        and canonicalize_version(p.version) == normalized_version
-        and p.package_type == "sdist"
-    ]
 
-    if not sdists:
-        raise PackageRejected(
-            f"No sdists found for package {name}=={version}",
-            solution=(
-                "It seems that this version does not exist or isn't published as a sdist "
-                "(a zip or a tarball).\n"
-                "You may be able to specify the dependency directly via a URL instead, "
-                "for example the tarball for a GitHub release."
-            ),
-            docs=PIP_NO_SDIST_DOC,
-        )
+        if dpi.package_type == "sdist":
+            sdists.append(dpi)
+        else:
+            if dpi.should_download_wheel():
+                wheels.append(dpi)
+            else:
+                log.info("Filtering out %s due to checksum mismatch", package.filename)
 
-    best_sdist = max(sdists, key=_sdist_preference)
-    if best_sdist.is_yanked:
-        raise PackageRejected(
-            f"All sdists for package {name}=={version} are yanked",
-            solution=(
-                f"Please update the {name} version in your requirements file.\n"
-                "Usually, when a version gets yanked from PyPI, there will already "
-                "be a fixed version available.\n"
-                "Otherwise, you may need to pin to the previous version."
-            ),
-        )
+    if len(sdists) != 0:
+        best_sdist = max(sdists, key=_sdist_preference)
+        if best_sdist.is_yanked:
+            raise PackageRejected(
+                f"All sdists for package {name}=={version} are yanked",
+                solution=(
+                    f"Please update the {name} version in your requirements file.\n"
+                    "Usually, when a version gets yanked from PyPI, there will already "
+                    "be a fixed version available.\n"
+                    "Otherwise, you may need to pin to the previous version."
+                ),
+            )
+    else:
+        log.warning("No source distributions found for package %s==%s", name, version)
+        best_sdist = None
+
+        if len(wheels) == 0:
+            if allow_binary:
+                solution = "Try to specify the dependency directly via a URL instead,\nfor example the tarball for a GitHub release."
+            else:
+                solution = "Please check that the package exists on PyPI.\nConsider using option allow_binary to allow wheels."
```

Comment:
IMO these are reversed

If allow_binary is true and there are 0 sdists and 0 wheels, the user probably just messed up the package name or version. There doesn't need to be any complicated solution or docs, just ask the user to double-check.

If allow_binary is false, then we're only looking for sdists, not wheels. That's very unusual from a user's perspective and we need to explain what went wrong. We should also include a link to the docs, like the original error message: https://github.com/containerbuildsystem/cachi2/pull/319/files?diff=split&w=0#diff-cfda53cfc3e45684cd3ac813162ad9ee144594345ff841a69250cc88de960dbcL1787-L1795 (I think we can keep the original error message)